File: hooks/scripts/post-write.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import shutil
import subprocess
import sys

sys.path.insert(0, os.path.dirname(__file__))
from _lib import (
    AI_PROSE_SLOP,
    COMMENT_SLOP_PATTERNS,
    PLACEHOLDER_HARD,
    deny,
    is_prose_file,
    is_test_file,
    passthrough,
    read_stdin,
    warn,
)

FORMATTERS: dict[str, list[list[str]]] = {
    "py": [["ruff", "format", "--quiet"], ["black", "--quiet"]],
    "go": [["gofmt", "-w"]],
    "rs": [["rustfmt", "--edition", "2021"]],
    "ts": [["prettier", "--write"], ["deno", "fmt"]],
    "tsx": [["prettier", "--write"], ["deno", "fmt"]],
    "js": [["prettier", "--write"], ["deno", "fmt"]],
    "jsx": [["prettier", "--write"], ["deno", "fmt"]],
    "swift": [["swift-format", "format", "--in-place"]],
    "cpp": [["clang-format", "-i"]],
    "cc": [["clang-format", "-i"]],
    "cxx": [["clang-format", "-i"]],
    "c": [["clang-format", "-i"]],
    "h": [["clang-format", "-i"]],
    "hpp": [["clang-format", "-i"]],
}


def get_tool_file_and_content(data: dict) -> tuple[str, str]:
    tool_name = data.get("tool_name", "")
    tool_input = data.get("tool_input", {})
    file_path = tool_input.get("file_path", "")
    if tool_name == "Write":
        content = tool_input.get("content", "")
    elif tool_name == "Edit":
        content = tool_input.get("new_string", "")
    else:
        content = ""
    return file_path, content


def run_formatter(file_path: str) -> str | None:
    ext = os.path.splitext(file_path)[1].lstrip(".")
    if not ext or not os.path.isfile(file_path):
        return None
    for cmd_parts in FORMATTERS.get(ext, []):
        if shutil.which(cmd_parts[0]):
            try:
                with open(file_path, "rb") as f:
                    before = f.read()
                subprocess.run(
                    [*cmd_parts, file_path],
                    capture_output=True,
                    timeout=30,
                )
                with open(file_path, "rb") as f:
                    after = f.read()
                if after != before:
                    return cmd_parts[0]
                return None
            except (subprocess.TimeoutExpired, OSError):
                pass
    return None


def placeholder_patterns(content: str, file_path: str) -> list[str]:
    if is_test_file(file_path):
        return []
    return [pat.pattern for pat in PLACEHOLDER_HARD if pat.findall(content)]


def slop_patterns(content: str, file_path: str) -> list[str]:
    hits: list[str] = []
    for pat in COMMENT_SLOP_PATTERNS:
        if pat.search(content):
            hits.append("narrating comment")
            break
    if is_prose_file(file_path):
        for pat in AI_PROSE_SLOP:
            if pat.search(content):
                hits.append(pat.pattern)
    return hits[:5]


def main() -> None:
    data = read_stdin()
    if not data:
        passthrough()
    file_path, content = get_tool_file_and_content(data)
    if not file_path:
        passthrough()
    formatter_used = run_formatter(file_path)
    format_note = (
        f" [format] File was auto-formatted by {formatter_used}. Your output was adjusted." if formatter_used else ""
    )
    if not content:
        if format_note:
            warn(format_note.strip())
        passthrough()
    placeholders = placeholder_patterns(content, file_path)
    if placeholders:
        deny(
            f"Placeholder code in {os.path.basename(file_path)}: "
            f"{', '.join(placeholders[:3])}. "
            f"Finish the implementation.{format_note}",
            event="PostToolUse",
        )
    slop = slop_patterns(content, file_path)
    if slop:
        warn(
            f"Comment/prose slop in {os.path.basename(file_path)} "
            f"({', '.join(slop[:3])}). "
            f"Remove narrating comments and AI filler.{format_note}"
        )
    if format_note:
        warn(format_note.strip())
    passthrough()


if __name__ == "__main__":
    main()

File: hooks/scripts/post-write.mjs
#!/usr/bin/env node
import { execFileSync, spawnSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { basename, extname } from "node:path";
import {
	AI_PROSE_SLOP,
	COMMENT_SLOP_PATTERNS,
	deny,
	isProseFile,
	isTestFile,
	PLACEHOLDER_HARD,
	passthrough,
	readStdin,
	SYCOPHANCY_PATTERNS,
	warn,
} from "./_lib.mjs";

const FORMATTERS = {
	py: [
		["ruff", "format", "--quiet"],
		["black", "--quiet"],
	],
	go: [["gofmt", "-w"]],
	rs: [["rustfmt", "--edition", "2021"]],
	ts: [
		["prettier", "--write"],
		["deno", "fmt"],
	],
	tsx: [
		["prettier", "--write"],
		["deno", "fmt"],
	],
	js: [
		["prettier", "--write"],
		["deno", "fmt"],
	],
	jsx: [
		["prettier", "--write"],
		["deno", "fmt"],
	],
	swift: [["swift-format", "format", "--in-place"]],
	cpp: [["clang-format", "-i"]],
	cc: [["clang-format", "-i"]],
	cxx: [["clang-format", "-i"]],
	c: [["clang-format", "-i"]],
	h: [["clang-format", "-i"]],
	hpp: [["clang-format", "-i"]],
};

function which(cmd) {
	try {
		execFileSync("which", [cmd], { stdio: "ignore" });
		return true;
	} catch {
		return false;
	}
}

function getToolFileAndContent(data) {
	const toolName = data.tool_name ?? "";
	const toolInput = data.tool_input ?? {};
	const filePath = toolInput.file_path ?? "";
	let content = "";
	if (toolName === "Write") content = toolInput.content ?? "";
	else if (toolName === "Edit") content = toolInput.new_string ?? "";
	return [filePath, content];
}

function runFormatter(filePath) {
	const ext = extname(filePath).replace(/^\./, "");
	if (!ext || !existsSync(filePath)) return null;
	const candidates = FORMATTERS[ext] ?? [];
	for (const cmdParts of candidates) {
		if (!which(cmdParts[0])) continue;
		try {
			const before = readFileSync(filePath);
			spawnSync(cmdParts[0], [...cmdParts.slice(1), filePath], {
				timeout: 30000,
			});
			const after = readFileSync(filePath);
			if (!before.equals(after)) return cmdParts[0];
			return null;
		} catch {
			// try next formatter
		}
	}
	return null;
}

function placeholderPatterns(content, filePath) {
	if (isTestFile(filePath)) return [];
	return PLACEHOLDER_HARD.filter((pat) => pat.test(content)).map(
		(pat) => pat.source,
	);
}

function slopPatterns(content) {
	const hits = [];
	if (COMMENT_SLOP_PATTERNS.some((pat) => pat.test(content))) {
		hits.push("narrating comment");
	}
	for (const pat of AI_PROSE_SLOP) {
		if (pat.test(content)) hits.push(pat.source);
	}
	return hits.slice(0, 5);
}

function sycophancyPatterns(content) {
	return SYCOPHANCY_PATTERNS.filter((pat) => pat.test(content))
		.map((pat) => pat.source)
		.slice(0, 5);
}

const data = readStdin();
if (!data) passthrough();

const [filePath, content] = getToolFileAndContent(data);
if (!filePath) passthrough();

const formatterUsed = runFormatter(filePath);
const formatNote = formatterUsed
	? ` [format] File was auto-formatted by ${formatterUsed}. Your output was adjusted.`
	: "";

if (!content) {
	if (formatNote) warn(formatNote.trim());
	passthrough();
}

const placeholders = placeholderPatterns(content, filePath);
if (placeholders.length > 0) {
	deny(
		`Placeholder code in ${basename(filePath)}: ` +
			`${placeholders.slice(0, 3).join(", ")}. ` +
			`Finish the implementation.${formatNote}`,
		"PostToolUse",
	);
}

const prose = isProseFile(filePath);
const slop = slopPatterns(content);
if (slop.length > 0) {
	const slopMsg =
		`Comment/prose slop in ${basename(filePath)} ` +
		`(${slop.slice(0, 3).join(", ")}). ` +
		`Remove narrating comments and AI filler.${formatNote}`;
	if (prose) warn(slopMsg);
	else deny(slopMsg, "PostToolUse");
}

if (prose) {
	const syco = sycophancyPatterns(content);
	if (syco.length > 0) {
		warn(
			`Sycophantic phrasing in ${basename(filePath)} ` +
				`(${syco.slice(0, 3).join(", ")}). ` +
				`Remove filler openers and apology phrases.${formatNote}`,
		);
	}
}

if (formatNote) warn(formatNote.trim());
passthrough();
